### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by aknysh**
## what
* Initial implementation

## why
*  Terraform module that accepts local or remote YAML stack configurations and returns deep-merged variables, backend config, and remote state outputs for Terraform and helmfile components.

The module is composed of three sub-modules:

  - [vars](modules/vars) - accepts stack configuration and returns deep-merged variables for a Terraform or helmfile component.
  - [backend](modules/backend) - accepts stack configuration and returns backend config for a Terraform component.
  - [remote-state](modules/remote-state) - accepts stack configuration and returns remote state outputs for a Terraform component. The module supports `s3` and `remote` (Terraform Cloud) backends.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,493 @@
...
+
+## Outputs
+
+| Name | Description |
+|------|-------------|
+| all\_imports\_list | List of all imported YAML configurations |
+| all\_imports\_map | Map of all imported YAML configurations |
+| backend | Backend configuration for a component |
+| backend\_type | Backend type configuration |
+| list\_configs | Terraform lists from YAML configurations |
```
Comment (osterman) [inline README.md:280]: I don't know what this is as it relates to stacks. Also do we define stacks and what they represent?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,493 @@
...
+| map\_configs | Terraform maps from YAML configurations |
```
Comment (osterman) [inline README.md:281]: Same as comment above

File: README.md (hunk 1 above)
Comment (osterman) [reply README.md:280]: Seems like we should have an output for all components but the word component isn't mentioned.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,493 @@
...
+
+```hcl
+module "yaml_stack_config" {
+  source = "cloudposse/stack-config/yaml"
+  # Cloud Posse recommends pinning every module to a specific version
+  # version     = "x.x.x"
+
+  map_config_local_base_path = "./config"
+
+  map_config_paths = [
```
Comment (osterman) [inline README.md:141]: Stack config paths, no? 

This module should be opinionated

File: README.yaml [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+  - name: "Terraform `template_file` data source"
+    description: "The `template_file` data source renders a template from a template string, which is usually loaded from an external file."
+    url: "https://registry.terraform.io/providers/hashicorp/template/latest/docs/data-sources/file"
+  - name: "Deepmerge"
+    description: "Terraform module to perform a deep map merge of standard Terraform maps and objects."
+    url: "https://github.com/Imperative-Systems-Inc/terraform-modules/tree/master/deepmerge"
+
+# Short description of this project
+description: |-
+  Terraform module to convert local and remote YAML configuration templates into deep-merged maps
```
Comment (osterman) [inline README.yaml:93]: This module is more about dealing with stacks. These are implementation details.

File: README.yaml [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+  of variables and Terraform backend attributes to provision [Terraform](https://www.terraform.io/) and [helmfile](https://github.com/roboll/helmfile) stacks.
+
+# Introduction to the project
+introduction: |-
+
```
Comment (osterman) [inline README.yaml:98]: Define stacks. Why should someone be interested in using stacks?

File: README.yaml [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    for the specified component type and component name
+
+  The module returns the following:
+
+  - List of all imported files
+  - The final deep merged stack configuration from all provided and imported YAML files
+  - The final configuration of all variables (`vars`) for the specified component type and component name
+  - The final configuration of all Terraform backend attributes (`backend`) for the specified component
+
+  ### Attributions
```
Comment (osterman) [inline README.yaml:134]: This module does not have the der merge module.

File: README.yaml [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+  For automated tests of the complete example using [bats](https://github.com/bats-core/bats-core) and [Terratest](https://github.com/gruntwork-io/terratest)
+  (which tests and deploys the example on Datadog), see [test](test).
+
+
+# Example usage
+examples: |-
+
+  ```hcl
+  module "yaml_stack_config" {
```
Comment (osterman) [inline README.yaml:152]:
```suggestion
  module "example_stack_config" {
```

File: README.yaml [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+    source = "cloudposse/stack-config/yaml"
+    # Cloud Posse recommends pinning every module to a specific version
+    # version     = "x.x.x"
+
+    map_config_local_base_path = "./config"
+
+    map_config_paths = [
```
Comment (osterman) [inline README.yaml:159]:
```suggestion
    stack_paths = [
```

File: README.yaml [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+# Example usage
+examples: |-
+
+  ```hcl
+  module "yaml_stack_config" {
+    source = "cloudposse/stack-config/yaml"
+    # Cloud Posse recommends pinning every module to a specific version
+    # version     = "x.x.x"
+
+    map_config_local_base_path = "./config"
```
Comment (osterman) [inline README.yaml:157]:
```suggestion
    stack_base_path = "./config"
```

File: README.yaml [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+      "my-stack.yaml"
+    ]
+
+    component_type = "terraform"
+    component      = "my-vpc"
+
+    context = module.this.context
+  }
+  ```
+
```
Comment (osterman) [inline README.yaml:169]: We also need to explain the structure of a stack configuration. 

We should also consider adding a `version` to the stack YAML configuration.

File: outputs.tf [hunk 11]
```diff
@@ -0,0 +1,34 @@
+output "map_configs" {
+  value       = module.yaml_config.map_configs
+  description = "Terraform maps from YAML configurations"
+}
+
+output "list_configs" {
+  value       = module.yaml_config.list_configs
+  description = "Terraform lists from YAML configurations"
+}
+
```
Comment (osterman) [inline outputs.tf:10]: We need to output components, complete with their vars.

File: outputs.tf [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+output "all_imports_list" {
+  value       = module.yaml_config.all_imports_list
+  description = "List of all imported YAML configurations"
+}
+
+output "all_imports_map" {
+  value       = module.yaml_config.all_imports_map
+  description = "Map of all imported YAML configurations"
+}
```
Comment (osterman) [inline outputs.tf:19]: What's the case for having the imports output as a map?

Comment (aknysh) [reply outputs.tf:19]: I don't know the exact use-case, but it shows some kind of a hierarchy - not sure if we need that

File: README.yaml [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+  - name: "Terraform `template_file` data source"
+    description: "The `template_file` data source renders a template from a template string, which is usually loaded from an external file."
+    url: "https://registry.terraform.io/providers/hashicorp/template/latest/docs/data-sources/file"
+  - name: "Deepmerge"
+    description: "Terraform module to perform a deep map merge of standard Terraform maps and objects."
+    url: "https://github.com/Imperative-Systems-Inc/terraform-modules/tree/master/deepmerge"
+
+# Short description of this project
+description: |-
+  Terraform module that accepts local or remote YAML stack configurations
```
Comment (osterman) [inline README.yaml:93]:
```suggestion
  Terraform module that loads an opinionated "stack" configuration from local or remote YAML sources.
```

File: README.yaml [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+  and returns deep-merged variables, backend config, and remote state outputs for Terraform and helmfile components.
```
Comment (osterman) [inline README.yaml:94]:
```suggestion
  It supports deep-merged variables, backend config, and remote state outputs for Terraform and helmfile components.
```

File: README.yaml [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+
+# Introduction to the project
+introduction: |-
+
+  The module has three sub-modules:
```
Comment (osterman) [inline README.yaml:99]:
```suggestion
  The module is composed of three sub-modules:
```

File: README.yaml [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+  For an example on how to configure remote state for Terraform components in YAML config files and then read the components outputs from the remote state,
+  see [examples/remote-state](examples/remote-state).
+
+  For automated tests of the complete example using [bats](https://github.com/bats-core/bats-core) and [Terratest](https://github.com/gruntwork-io/terratest),
+  see [test](test).
+
+
+# Example usage
+examples: |-
+
```
Comment (osterman) [inline README.yaml:125]: Here's an example stack configuration file:


```yaml
import:
    - ue2-globals

  vars:
    stage: dev

  terraform:
    vars: {}

  helmfile:
    vars: {}

  components:
    terraform:
      vpc:
        command: "/usr/bin/terraform-0.13"
        backend:
          s3:
            workspace_key_prefix: "vpc"
        vars:
          cidr_block: "10.102.0.0/18"
      eks:
        backend:
          s3:
            workspace_key_prefix: "eks"
        vars:

    helmfile:
      nginx-ingress:
        vars:
          installed: true
```

The `import` section refers to other stack configurations that are automatically deep merged.

**Review by osterman — APPROVED** [review]

**Pull request merged**